docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,754 @@
+and `oe_verify_evidence` respectively. These align with the ones proposed in
+the [endorsements design doc](RemoteAttestatonCollaterals.md). There will also
+be functions for registering and unregistering plugins called
+`oe_register_attestation_plugin` and `oe_unregister_attestation_plugin`. The user
+can statically link in their desired plugin and call the register plugin function.
+The attestation data can be retrieved from the "plugin aware" analogue of
+`oe_get_report` with the desired UUID. The generated data will have the UUID in
+its header. The user can call the analogue of `oe_verify_report` to verify the
+data and the Open Enclave runtime can use this UUID to determine what plugin
+verification routine to run.
+
+### Plugin API
+
+Each plugin must implement the functions below:
+
+```C
+/**
+ * Claims struct used for claims parameters for the plugin.
+ */
+struct oe_claim_t
+{
+    char* name;
+    uint8_t* value;
+    size_t value_size;
+};
+
+/**
+ * Struct that defines the structure of each plugin. Each plugin must
+ * define an UUID for its format and implement the functions in this
+ * struct. Ideally, each plugin should provide a helper function to
+ * create this struct on the behalf of the plugin users.
+ */
+struct oe_attestation_plugin_t
+{
+    /**
+     * The UUID for the plugin.
+     */
+    uuid_t format_id;
+
+    /**
+     * The function that gets executed when a plugin is registered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] config_data An optional pointer to the configuration data.
+     * @param[in] config_data_size The size in bytes of config_data.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_register)(
+        oe_attestation_plugin_t* plugin_context,
+        const void* config_data,
+        size_t config_data_size);
+
+    /**
+     * The function that gets executed when a plugin is unregistered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_unregister)(
+        oe_attestation_plugin_t* plugin_context);
+
+    /**
+     * Generates the attestation evidence, which is defined as the data
+     * produced by the enclave. The caller may pass in custom claims, which
+     * must be attached to the evidence and then cryptographically signed.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] flags Specifying default value (0) generates evidence for local
+     * attestation. Specifying OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION generates
+     * evidence for remote attestation.
+     * @param[in] custom_claims The optional custom claims list.
+     * @param[in] custom_claims_length The number of custom claims.
+     * @param[in] opt_params The optional plugin-specific input parameters.
+     * @param[in] opt_params_size The size of opt_params.
```

Comment:
"… in bytes." (same for lines 147 and 151)  For consistency with line 195, 197, etc.